### PR TITLE
Syntax highlighting that plays well with font-lock

### PR DESCRIPTION
**Pull request by ubolonton**
This is a continuation of #16.

- [x] Add a minor mode `tree-sitter-hl-mode`.
- [x] Add mechanism for major modes to customize syntax highlighting: `tree-sitter-hl-default-patterns`.
- [ ] ~Add default syntax highlighting patterns to the language bundle `tree-sitter-langs`.~  This will be left for later, after improving the binary distribution mechanism.
- [x] Face mapping
    + [x] Define faces for most common capture names (scopes) defined in the language repos, e.g. `@function.method` would map to `tree-sitter-hl-face:function.method`.
    + [x] Prioritize captures in earlier patterns, when multiple patterns match.
    + [x] Allow end user to add more query patterns: `tree-sitter-hl-add-patterns`.
    + [ ] ~Map capture names to face names automatically. For example `@:font-lock-keyword-face` can be mapped to the face `font-lock-keyword-face`.~ This will be done later, after discussing with upstream tree-sitter about standardization/conventions.
    + [x] Allow end user to selectively disable certain capture names.
- [ ] ~Respect custom keywords added by user through `font-lock-add-keywords`.~ This is too difficult currently, even for `font-lock-mode` itself (see `font-lock-refresh-defaults`).
- [x] Support code blocks in `org-mode` and `markdown-mode`.
- [ ] ~Support embedded blocks of other languages (injections).~ This will be left for later.

For now, trying this requires cloning the language repos into the directory `repos/` next to `tree-sitter-langs.el` (regardless of installation method).

I've been using this, and haven't noticed any performance difference from `font-lock-mode`.
In fact, sometimes it feels **faster**, in large C buffers.

### CODE REVIEW

File: lisp/tree-sitter.el [hunk 1]
```diff
@@ -104,46 +121,76 @@ END is the end of the changed text."
...
-(defun tree-sitter--disable ()
+(defun tree-sitter--teardown ()
   "Disable `tree-sitter' in the current buffer."
   (remove-hook 'after-change-functions #'tree-sitter--after-change :local)
   (remove-hook 'before-change-functions #'tree-sitter--before-change :local)
   (setq tree-sitter-tree nil
         tree-sitter-parser nil
         tree-sitter-language nil))
 
+(defmacro tree-sitter--error-protect (body-form &rest error-forms)
```
Comment (dgutov) [inline lisp/tree-sitter.el:146]: Maybe you want to use `condition-case` instead?

Comment (ubolonton) [reply lisp/tree-sitter.el:146]: This doesn't handle the error. It runs some cleanup code and lets the error propagate.

The closest `condition-case` equivalent I can think of doesn't preserve the backtrace:

```emacs-lisp
`(condition-case err
    ,body-form
   (error
    ,@error-forms
    (signal (car err) (cdr err))))
```

**Comment by ubolonton** [comment]
Extra query patterns can now be added dynamically:

```emacs-lisp
(add-hook 'rust-mode-hook
          (lambda ()
            (tree-sitter-hl-add-patterns
             [(call_expression
               function: (field_expression
                          field: (field_identifier) @function.call))
              (let_declaration (identifier) @variable)
              "?" @keyword
              (self) @keyword])))
```

Before:
rust default
After:
rust custom

**Comment by shackra** [comment]
amazing! didn't keep an eye on the project and suddenly, BANG! Syntax highlight is mostly done.

**Pull request merged**